mesh/spacing.hpp

```diff
@@ -0,0 +1,689 @@
+   {
+      // SPACING_TYPE numIntParam numDoubleParam {int params} {double params}
+      os << GAUSSIAN << " 3 2 " << n << " " << (int) reverse << " "
+         << (int) scale << " " << s0 << " " << s1 << "\n";
+   }
+
+   virtual SPACING_TYPE SpacingType() const override { return GAUSSIAN; }
+   virtual int NumIntParameters() const override { return 3; }
+   virtual int NumDoubleParameters() const override { return 2; }
+
+   virtual void GetIntParameters(Array<int> & p) const override
+   {
+      p.SetSize(3);
+      p[0] = n;
+      p[1] = (int) reverse;
+      p[2] = (int) scale;
+   }
+
+   virtual void GetDoubleParameters(Vector & p) const override
+   {
+      p.SetSize(2);
+      p[0] = s0;
+      p[1] = s1;
+   }
+
+   virtual bool Nested() const override { return false; }
+
+   virtual SpacingFunction *Clone() const override
+   {
+      return new GaussianSpacingFunction(n, reverse, s0, s1, scale);
+   }
+
+private:
+   double s0, s1; // First and last interval widths
+   Vector s;
+
+   void CalculateSpacing();
+};
+
+/** @brief Logarithmic spacing function, uniform in log base 10 by default.
+
+    The log base can be changed as an input parameter. Decreasing it makes the
+    distribution more uniform, whereas increasing it makes the spacing vary
+    more. Another input option is a flag to make the distribution symmetric
+    (default is non-symmetric). There are no scaled parameters. This function is
+    nested. */
+class LogarithmicSpacingFunction : public SpacingFunction
+{
+public:
+   LogarithmicSpacingFunction(int n_, bool r_, bool sym_=false, double b_=10.0)
+      : SpacingFunction(n_, r_), sym(sym_), logBase(b_)
+   {
+      CalculateSpacing();
+   }
+
+   virtual void SetSize(int size) override
+   {
+      n = size;
+      CalculateSpacing();
+   }
+
+   virtual double Eval(int p) override
+   {
+      const int i = reverse ? n - 1 - p : p;
+      return s[i];
+   }
+
+   virtual void Print(std::ostream &os) const override
+   {
+      // SPACING_TYPE numIntParam numDoubleParam {int params} {double params}
+      os << LOGARITHMIC << " 3 1 " << n << " " << (int) reverse << " "
+         << (int) sym << " " << logBase << "\n";
+   }
+
+   virtual SPACING_TYPE SpacingType() const override { return LOGARITHMIC; }
+   virtual int NumIntParameters() const override { return 3; }
+   virtual int NumDoubleParameters() const override { return 1; }
+
+   virtual void GetIntParameters(Array<int> & p) const override
+   {
+      p.SetSize(3);
+      p[0] = n;
+      p[1] = (int) reverse;
+      p[2] = (int) sym;
+   }
+
+   virtual void GetDoubleParameters(Vector & p) const override
+   {
+      p.SetSize(1);
+      p[0] = logBase;
+   }
+
+   virtual bool Nested() const override { return true; }
+
+   virtual SpacingFunction *Clone() const override
+   {
+      return new LogarithmicSpacingFunction(n, reverse, sym, logBase);
+   }
+
+private:
+   bool sym;
+   double logBase;
+   Vector s;
```

Comment:
Not sure about mfem's policy on doxygen of member variables, but can use `///< doc string here`, to doxygen small variables without increasing the line count.

Reply:
Done.